src/Collectors/PublicPropertyCollector.php

```diff
@@ -75,13 +77,40 @@ public function processNode(Node $node, Scope $scope): ?array
                     continue;
                 }
 
-                $publicPropertyNames[] = [$classReflection->getName(), $propertyName, $node->getLine()];
+                $isInternal = $this->isInternalRecursive($classReflection, $classLike, $propertyName, $scope);
+
+                $publicPropertyNames[] = [$classReflection->getName(), $propertyName, $node->getLine(), $isInternal];
             }
         }
 
         return $publicPropertyNames;
     }
 
+    private function isInternalRecursive(ClassReflection $classReflection, Class_ $class, string $propertyName, Scope $scope): bool
+    {
+        if ($this->internalDocStmtAnalyzer->isInternalDoc($class, $classReflection)) {
+            return true;
+        }
+
+        while (true) {
+            if (! $classReflection->hasProperty($propertyName)) {
+                return false;
+            }
+
+            $propertyReflection = $classReflection->getProperty($propertyName, $scope);
+
+            $docComment = $propertyReflection->getDocComment();
+            if ($docComment !== null && $this->internalDocStmtAnalyzer->isInternalDocComment($docComment)) {
+                return true;
+            }
+
+            $classReflection = $classReflection->getParentClass();
+            if (! $classReflection instanceof ClassReflection) {
+                return false;
+            }
+        }
+    }
```

Comment:
The `@internal` check must happen only on the element, to avoid enforcing by inheritance.
We do the same for `@api` by convention.